protobuf/src/test/java/io/grpc/protobuf/ProtoUtilsTest.java

```diff
@@ -86,4 +89,29 @@ public void marshallerShouldNotLimitProtoSize() throws Exception {
     // Just perform a round trip to verify that it works.
     testRoundtrip();
   }
+
+  @Test
+  public void testAvailable() throws Exception {
+    InputStream is = marshaller.stream(proto);
+    assertEquals(proto.getSerializedSize(), is.available());
+    is.read();
+    assertEquals(proto.getSerializedSize() - 1, is.available());
+    while (is.read() != -1) {}
+    assertEquals(-1, is.read());
+    assertEquals(0, is.available());
+  }
+
+  @Test
+  public void testEmpty() throws IOException {
+    Marshaller<Empty> marshaller = ProtoUtils.marshaller(Empty.getDefaultInstance());
+    InputStream is = marshaller.stream(Empty.getDefaultInstance());
+    assertEquals(0, is.available());
+    byte[] b = new byte[10];
+    assertEquals(-1, is.read(b));
+    assertArrayEquals(new byte[10], b);
+    assertEquals(-1, is.read(b));
+    assertArrayEquals(new byte[10], b);
```

Comment:
Yes, but that is on purpose. It is a pretty common bug where read()'ing again after the initial EOF produces an incorrect result. Since this is white-box testing, I also know that the internal state would be different between the two calls.

I added a comment.